### PR TITLE
Synchronize to June 3, 2016 of btcd

### PR DESCRIPTION
**Pull request by cjepson**
This is a bundle of iteratively merged commits onto master of mainly sigcache and memory optimizations that also apply to dcrd.  This should dramatically reduce the number of memory allocations when reading transactions on the wire, which should help dcrd sync faster and more efficiently under the new ticket DB code.

### CODE REVIEW

File: chaincfg/params.go [hunk 1]
```diff
@@ -318,7 +318,6 @@ var MainNetParams = Params{
 		{24480, newHashFromStr("0000000000000c9d4239c4ef7ef3fb5aaeed940244bc69c57c8c5e1f071b28a6")},
 		{48590, newHashFromStr("0000000000000d5e0de21a96d3c965f5f2db2c82612acd7389c140c9afe92ba7")},
 		{54770, newHashFromStr("00000000000009293d067b1126b7de07fc9b2b94ee50dfe0d48c239a7adb072c")},
-		{60720, newHashFromStr("0000000000000a64475d68ffb9ad89a3d147c0f5138db26b40da9d19d0004117")},
```
Comment (jolan) [inline chaincfg/params.go]: These were just added in PR #348.

File: mining_test.go [hunk 2]
```diff
@@ -1,5 +1,5 @@
 // Copyright (c) 2015-2016 The Decred Developers
-// Copyright (c) 2015-2016 The Decred developers
+// Copyright (c) 2016 The btcsuite developers
 // Use of this source code is governed by an ISC
```
Comment (jolan) [inline mining_test.go]: Should be like txscript/sigcache.go:

```
// Copyright (c) 2015-2016 The btcsuite developers
// Copyright (c) 2016 The Decred developers
```

**Comment by cjepson** [comment]
@jolan Those changes were reverted, should be OK now.

**Comment by davecgh** [comment]
I highly recommend against doing big syncs like this.  Do it a commit or two a time so they can be properly reviewed.

**Comment by cjepson** [comment]
@davecgh They were done a commit at a time. I was informed that I had to do a merge rather than a cherry-pick, so I adjusted the PR branch accordingly. The individual commits may be seen here:

https://github.com/cjepson/btcd/tree/cj_200916sync2

**Comment by cjepson** [comment]
This will be resubmitted as individual merges.

**Pull request closed without merging**